src/Polysemy/Final/Internal.hs

```diff
@@ -0,0 +1,211 @@
+                 -> (forall x. f x -> Maybe x)
+                 -> m (f a)
+                 )
+              -> Final m z a
+
+makeSem_ ''Final
+
+-----------------------------------------------------------------------------
+-- | Allows for embedding higher-order actions of the final monad
+-- by providing the means of explicitly threading effects through @'Sem' r@
+-- to the final monad.
+--
+-- Consider using 'Polysemy.Final.withStrategic' instead,
+-- which provides a more user-friendly interface, but is also slightly weaker.
+--
+-- You are discouraged from using 'withWeaving' directly in application code,
+-- as it ties your application code directly to the final monad.
+withWeaving :: forall m a r
+             . Member (Final m) r
+            => (  forall f
+                . Functor f
+               => f ()
+               -> (forall x. f (Sem r x) -> m (f x))
+               -> (forall x. f x -> Maybe x)
+               -> m (f a)
+               )
+            -> Sem r a
+
+data Strategy m f n z a where
+  GetInitialState     :: Strategy m f n z (f ())
+  HoistInterpretation :: (a -> n b) -> Strategy m f n z (f a -> m (f b))
+  GetInspector        :: Strategy m f n z (Inspector f)
+
+------------------------------------------------------------------------------
+-- | 'Strategic' is an environment in which you're capable of explicitly
+-- threading higher-order effect states to the final monad.
+-- This is a variant of @Tactics@ (see 'Tactical'), and usage
+-- is extremely similar.
+type Strategic m n a = forall f. Functor f => Sem (WithStrategy m f n) (m (f a))
+
+type WithStrategy m f n = '[Strategy m f n]
+
+------------------------------------------------------------------------------
+-- | Internal function to process Strategies in terms of 'withWeaving'.
+runStrategy :: Functor f
+            => f ()
+            -> (forall x. f (n x) -> m (f x))
+            -> (forall x. f x -> Maybe x)
+            -> Sem '[Strategy m f n] a
+            -> a
+runStrategy s wv ins = (run .) $ interpret $ \case
+  GetInitialState       -> pure s
+  HoistInterpretation f -> pure $ \fa -> wv (f <$> fa)
+  GetInspector          -> pure (Inspector ins)
+{-# INLINE runStrategy #-}
+
+
+------------------------------------------------------------------------------
+-- | Like 'interpretFinal' specialized to 'IO', but also tries very hard
+-- to preserve state semantics dependant on the order interpreters are run,
+-- adressing the primary issue with 'Final'.
+--
+-- Semantically, interpreters written using this behave very much as
+-- though they were written using 'withLowerToIO'.
+-- However, this does not need to spawn an interpreter thread, making
+-- it more efficient (but not any more safe.)
+--
+-- 'interpretFinalGlobal' operates under the assumption that any effectful
+-- state which can't be inspected using 'Polysemy.Inspector' can't contain any
+-- values. This is true for all interpreters featured in polysemy,
```

Comment:
I'n not sure what it means to not contain any values. The functorial state for `State` contains values, doesn't it?

Reply:
Yeah, but the functorial state for `State` is always inspectable. Compare this to the functorial state for `Error`: `Either e a`. If it's uninspectable, then it must be a `Left` value, and thus it doesn't contain any values of `a`.

Reply:
I understand, thanks! Mind adding that description to the comment?

Reply:
Sure thing.